diff --git a/src/bpf_assembler.cc b/src/bpf_assembler.cc
index 23cead1..55dd049 100644
--- a/src/bpf_assembler.cc
+++ b/src/bpf_assembler.cc
@@ -187,6 +187,12 @@ typedef class _bpf_assembler
             inst.opcode = EBPF_OP_LDXH;
         } else if (mnemonic == "ldxw") {
             inst.opcode = EBPF_OP_LDXW;
+        } else if (mnemonic == "ldxbsx") {
+            inst.opcode = EBPF_OP_LDXBSX;
+        } else if (mnemonic == "ldxhsx") {
+            inst.opcode = EBPF_OP_LDXHSX;
+        } else if (mnemonic == "ldxwsx") {
+            inst.opcode = EBPF_OP_LDXWSX;
         }
 
         return inst;
@@ -486,8 +492,10 @@ typedef class _bpf_assembler
         {"jsle", {&_bpf_assembler::_encode_jmp, 3}},  {"jsle32", {&_bpf_assembler::_encode_jmp, 3}},
         {"jslt", {&_bpf_assembler::_encode_jmp, 3}},  {"jslt32", {&_bpf_assembler::_encode_jmp, 3}},
         {"lddw", {&_bpf_assembler::_encode_ld, 2}},   {"ldxb", {&_bpf_assembler::_encode_ldx, 2}},
-        {"ldxdw", {&_bpf_assembler::_encode_ldx, 2}}, {"ldxh", {&_bpf_assembler::_encode_ldx, 2}},
-        {"ldxw", {&_bpf_assembler::_encode_ldx, 2}},  {"le16", {&_bpf_assembler::_encode_alu, 1}},
+        {"ldxbsx", {&_bpf_assembler::_encode_ldx, 2}}, {"ldxdw", {&_bpf_assembler::_encode_ldx, 2}},
+        {"ldxh", {&_bpf_assembler::_encode_ldx, 2}},   {"ldxhsx", {&_bpf_assembler::_encode_ldx, 2}},
+        {"ldxw", {&_bpf_assembler::_encode_ldx, 2}},   {"ldxwsx", {&_bpf_assembler::_encode_ldx, 2}},
+        {"le16", {&_bpf_assembler::_encode_alu, 1}},
         {"le32", {&_bpf_assembler::_encode_alu, 1}},  {"le64", {&_bpf_assembler::_encode_alu, 1}},
         {"lsh", {&_bpf_assembler::_encode_alu, 2}},   {"lsh32", {&_bpf_assembler::_encode_alu, 2}},
         {"mod", {&_bpf_assembler::_encode_alu, 2}},   {"mod32", {&_bpf_assembler::_encode_alu, 2}},
diff --git a/src/ebpf.h b/src/ebpf.h
index 9ee52ca..a754a87 100644
--- a/src/ebpf.h
+++ b/src/ebpf.h
@@ -66,6 +66,7 @@ typedef struct ebpf_inst
 /* Other memory modes are not yet supported */
 #define EBPF_MODE_IMM 0x00
 #define EBPF_MODE_MEM 0x60
+#define EBPF_MODE_MEMSX 0x80
 
 #define EBPF_ALU_OP_ADD 0x00
 #define EBPF_ALU_OP_SUB 0x10
@@ -141,6 +142,9 @@ typedef struct ebpf_inst
 #define EBPF_OP_LDXH (EBPF_CLS_LDX | EBPF_MODE_MEM | EBPF_SIZE_H)
 #define EBPF_OP_LDXB (EBPF_CLS_LDX | EBPF_MODE_MEM | EBPF_SIZE_B)
 #define EBPF_OP_LDXDW (EBPF_CLS_LDX | EBPF_MODE_MEM | EBPF_SIZE_DW)
+#define EBPF_OP_LDXWSX (EBPF_CLS_LDX | EBPF_MODE_MEMSX | EBPF_SIZE_W)
+#define EBPF_OP_LDXHSX (EBPF_CLS_LDX | EBPF_MODE_MEMSX | EBPF_SIZE_H)
+#define EBPF_OP_LDXBSX (EBPF_CLS_LDX | EBPF_MODE_MEMSX | EBPF_SIZE_B)
 #define EBPF_OP_STW (EBPF_CLS_ST | EBPF_MODE_MEM | EBPF_SIZE_W)
 #define EBPF_OP_STH (EBPF_CLS_ST | EBPF_MODE_MEM | EBPF_SIZE_H)
 #define EBPF_OP_STB (EBPF_CLS_ST | EBPF_MODE_MEM | EBPF_SIZE_B)
